src/operator/aggregate.rs

```diff
@@ -178,19 +183,32 @@ where
     }
 }
 
-impl<I, F, O> UnaryOperator<I, O> for Aggregate<I, F, O>
+impl<Z, F, O> UnaryOperator<Z, O> for Aggregate<Z, F, O>
 where
-    I: BatchReader<R = O::R> + 'static,
-    F: Fn(&I, &mut I::Cursor) -> O::Key + 'static,
+    Z: IndexedZSet<R = O::R> + 'static,
+    F: Fn(&Z::Key, &mut [(&Z::Val, Z::R)]) -> O::Key + 'static,
     O: Clone + ZSet + 'static,
     O::R: ZRingValue,
 {
-    fn eval(&mut self, i: &I) -> O {
+    fn eval(&mut self, i: &Z) -> O {
         let mut elements = Vec::with_capacity(i.len());
         let mut cursor = i.cursor();
+        let mut vals: Vec<(&Z::Val, Z::R)> = Vec::new();
```

Comment:
Should we pre-allocate this with the number of keys or values?

Reply:
At this point we have no insight about the average/max size of a group. But I guess there's not much harm in preallocating `delta.len()` elements.